functests/dpdk/dpdk.go

```diff
@@ -997,14 +1019,51 @@ func waitForSRIOVStable() {
 	}, waitingTime, 1*time.Second).Should(BeTrue())
 }
 
-func filterPerformanceProfilesWithAvailableNodes(performanceProfiles []*perfv1alpha1.PerformanceProfile, nodeSelector map[string]string) []*perfv1alpha1.PerformanceProfile {
-	availableProfiles := make([]*perfv1alpha1.PerformanceProfile, 0)
+// findPerformanceProfileAndPolicyWithAvailableNodes finds a profile/sriovPolicy match for which a node with
+// allocatable resources is available. It will return a profile/sriovPolicy for a policy with resource name
+// "dpdknic", or a pair with the most available resource on node
+func findPerformanceProfileAndPolicyWithAvailableNodes(performanceProfiles []*perfv1alpha1.PerformanceProfile, nodeSelector map[string]string,
+) (preferredProfile *perfv1alpha1.PerformanceProfile, preferredResource string, preferredDevice *sriovv1.InterfaceExt, err error) {
+	currentResourceCount := 0
+	sriovInfos, err := sriovcluster.DiscoverSriov(sriovclient, SRIOV_OPERATOR_NAMESPACE)
+	Expect(err).ToNot(HaveOccurred())
+
+	sriovPolicies := &sriovv1.SriovNetworkNodePolicyList{}
+	err = client.Client.List(context.TODO(), sriovPolicies, &goclient.ListOptions{Namespace: SRIOV_OPERATOR_NAMESPACE})
+	Expect(err).ToNot(HaveOccurred())
+
 	for _, profile := range performanceProfiles {
 		profileNodeSelector := nodes.NodesSelectorUnion(nodeSelector, profile.Spec.NodeSelector)
-		nodesAvailable := nodes.ValidateNodeIsAvailableForSelector(profileNodeSelector)
-		if nodesAvailable {
-			availableProfiles = append(availableProfiles, profile)
+		nodesAvailable, err := nodes.NodesAvailableForSelector(profileNodeSelector)
+		Expect(err).ToNot(HaveOccurred())
+
+		for _, sriovPolicy := range sriovPolicies.Items {
+			if sriovPolicy.Spec.DeviceType != "netdevice" {
+				continue
+			}
+			if &sriovPolicy.Spec.NicSelector == nil || &sriovPolicy.Spec.NicSelector.PfNames == nil || len(sriovPolicy.Spec.NicSelector.PfNames) != 1 {
+				continue
+			}
+			for _, node := range nodesAvailable {
+				quantity := node.Status.Allocatable[corev1.ResourceName("openshift.io/"+sriovPolicy.Spec.ResourceName)]
+				resourceCount64, _ := (&quantity).AsInt64()
+				resourceCount := int(resourceCount64)
+				sriovDevice, err := sriovInfos.FindOneSriovDevice(node.ObjectMeta.Name)
+				if err != nil {
+					continue
+				}
+				// Return profile and policy with the prefered resource name if available
+				if sriovPolicy.Spec.ResourceName == resourceName {
+					return profile, sriovPolicy.Spec.ResourceName, sriovDevice, nil
```

Comment:
Can you be consistent with named vs non named returns?
Probably it comes more natural if we return a struct.

Reply:
Done